src/cpu/testers/spatter_gen/spatter_gen.hh

```diff
@@ -0,0 +1,239 @@
+/*
+* Copyright (c) 2024 The Regents of The University of California
+* All rights reserved.
+*
+* Redistribution and use in source and binary forms, with or without
+* modification, are permitted provided that the following conditions are
+* met: redistributions of source code must retain the above copyright
+* notice, this list of conditions and the following disclaimer;
+* redistributions in binary form must reproduce the above copyright
+* notice, this list of conditions and the following disclaimer in the
+* documentation and/or other materials provided with the distribution;
+* neither the name of the copyright holders nor the names of its
+* contributors may be used to endorse or promote products derived from
+* this software without specific prior written permission.
+*
+* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+* "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+* LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+* A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+* OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+* SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+* LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+* DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+* THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+* (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+* OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+#ifndef __CPU_TESTERS_SPATTER_GEN_SPATTER_GEN_HH__
+#define __CPU_TESTERS_SPATTER_GEN_SPATTER_GEN_HH__
+
+#include <queue>
+#include <unordered_map>
+#include <vector>
+
+#include "base/statistics.hh"
+#include "base/stats/group.hh"
+#include "cpu/testers/spatter_gen/spatter_kernel.hh"
+#include "enums/SpatterKernelType.hh"
+#include "mem/packet.hh"
+#include "mem/port.hh"
+#include "params/SpatterGen.hh"
+#include "sim/clocked_object.hh"
+#include "sim/eventq.hh"
+
+namespace gem5
+{
+
+typedef std::tuple<PacketPtr, Tick> TimedPacket;
+
+PacketPtr entryPacket(const TimedPacket& timed_pkt);
+Tick entryTick(const TimedPacket& timed_pkt);
+
+using enums::SpatterKernelType;
+
+class SpatterGen: public ClockedObject
+{
+  private:
+
+    class SpatterGenEvent : public EventFunctionWrapper
+    {
+      // TODO: split pending into pendingInput and pendingOutput
+      private:
+        enum SleepState
+        {
+            AWAKE,
+            ASLEEP
+        };
+
+        SleepState _state;
```

Comment:
Seems like this could just be a boolean `isPending` (or `isWaiting` or something like that)

Reply:
I like to keep it an enum, I know right now the state is a boolean like value. In future I want to extend this so that it can encode more states, e.g. ASLEEP_ON_INPUT, ASLEEP_ON_OUTPUT, ...